modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/LimitOffsetTest.java

```
@@ -0,0 +1,154 @@
+            .setCacheMode(CacheMode.PARTITIONED)
+            .setBackups(1)
+            .setQueryEntities(singletonList(ent))
+            .setSqlSchema("PUBLIC");
+    }
+
+    /**
+     *
+     */
+    @Test
+    public void testLimitOffset() {
+        int[] limits = {-1, 0, 10, ROWS / 2 - 1, ROWS / 2, ROWS / 2 + 1, ROWS - 1, ROWS};
+        int[] offsets = {-1, 0, 10, ROWS / 2 - 1, ROWS / 2, ROWS / 2 + 1, ROWS - 1, ROWS};
+
+        for (int lim : limits) {
+            for (int off : offsets) {
+                log.info("+++ Check [limit=" + lim + ", off=" + off + ']');
+
+                checkQuery(lim, off, false, false);
+                checkQuery(lim, off, true, false);
+                checkQuery(lim, off, false, true);
+                checkQuery(lim, off, true, true);
+            }
+        }
+    }
+
+    /**
+     * Check query with specified limit and offset (or without its when the arguments are negative),
+     *
+     * @param lim Limit.
+     * @param off Offset.
+     * @param param If {@code false} place limit/offset as literals. Otherwise they are plase as parameters.
+     * @param sorted Use sorted query (adds ORDER BY).
+     */
+    void checkQuery(int lim, int off, boolean param, boolean sorted) {
+        QueryEngine engine = Commons.lookupComponent(grid(0).context(), QueryEngine.class);
+
+        String sql = createSql(lim, off, false, false);
+
+        log.info("SQL: " + sql);
+
+        List<FieldsQueryCursor<List<?>>> cursors =
+            engine.query(null, "PUBLIC", sql, param ? new Object[]{off, lim} : X.EMPTY_OBJECT_ARRAY);
+
+        List<List<?>> res = cursors.get(0).getAll();
+
+        assertEquals("Invalid results size. [limit=" + lim + ", off=" + off + ", res=" + res.size() + ']',
+            expectedSize(lim, off), res.size());
+    }
+
+    /**
+     * Calculates expected result set size by limit and offset.
+     */
+    private int expectedSize(int lim, int off) {
+        if (off < 0)
```

Should we throw an exception when limit or offset is negative? Or it's a valid value?

---

At the test i use negative value not use limit/offset in the query (see #createSql). 
I'll add new test to check negative limit offset